akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,506 @@
+
+        val newUnconfirmed =
+          if (supportResend) stateAfterAck.unconfirmed
+          else Vector.empty
+
+        if ((viaTimeout || newConfirmedSeqNr == s.firstSeqNr) && supportResend) {
+          // the last message was lost and no more message was sent that would trigger Resend
+          resendUnconfirmed(newUnconfirmed)
+        }
+
+        // when supportResend=false the requestedSeqNr window must be expanded if all sent messages were lost
+        val newRequestedSeqNr2 =
+          if (!supportResend && newRequestedSeqNr <= stateAfterAck.currentSeqNr)
+            stateAfterAck.currentSeqNr + (newRequestedSeqNr - newConfirmedSeqNr)
+          else
+            newRequestedSeqNr
+        if (newRequestedSeqNr2 != newRequestedSeqNr)
+          ctx.log.infoN(
+            "Expanded requestedSeqNr from [{}] to [{}], because current [{}] and all were probably lost",
+            newRequestedSeqNr,
+            newRequestedSeqNr2,
+            stateAfterAck.currentSeqNr)
+
+        if (newRequestedSeqNr2 > s.requestedSeqNr) {
+          if (!s.requested && (newRequestedSeqNr2 - s.currentSeqNr) > 0)
+            s.producer ! RequestNext(producerId, s.currentSeqNr, newConfirmedSeqNr, msgAdapter, ctx.self)
+          active(
+            stateAfterAck.copy(
+              requested = true,
+              requestedSeqNr = newRequestedSeqNr2,
+              supportResend = supportResend,
+              unconfirmed = newUnconfirmed))
+        } else {
+          active(stateAfterAck.copy(supportResend = supportResend, unconfirmed = newUnconfirmed))
+        }
+
+      case Ack(newConfirmedSeqNr) =>
+        ctx.log.infoN("Ack, confirmed [{}], current [{}]", newConfirmedSeqNr, s.currentSeqNr)
+        val stateAfterAck = onAck(newConfirmedSeqNr)
+        if (newConfirmedSeqNr == s.firstSeqNr && stateAfterAck.unconfirmed.nonEmpty) {
+          resendUnconfirmed(stateAfterAck.unconfirmed)
+        }
+        active(stateAfterAck)
+
+      case Resend(fromSeqNr) =>
+        val newUnconfirmed = s.unconfirmed.dropWhile(_.seqNr < fromSeqNr)
+        resendUnconfirmed(newUnconfirmed)
+        active(s.copy(unconfirmed = newUnconfirmed))
+
+      case ResendFirst =>
+        if (s.unconfirmed.nonEmpty && s.unconfirmed.head.seqNr == s.firstSeqNr) {
+          ctx.log.info("resending first, [{}]", s.firstSeqNr)
+          s.send(s.unconfirmed.head.copy(first = true)(ctx.self))
+        } else {
+          if (s.currentSeqNr > s.firstSeqNr)
+            timers.cancel(ResendFirst)
+        }
+        Behaviors.same
+
+      case start: Start[A] @unchecked =>
+        ctx.log.info("Register new Producer [{}], currentSeqNr [{}].", start.producer, s.currentSeqNr)
+        if (s.requested)
+          start.producer ! RequestNext(producerId, s.currentSeqNr, s.confirmedSeqNr, msgAdapter, ctx.self)
+        active(s.copy(producer = start.producer))
+
+      case RegisterConsumer(consumerController: ActorRef[ConsumerController.Command[A]] @unchecked) =>
```

Comment:
Not that I love long class names but perhaps RegisterConsumerController would be worth it here since there also is a Consumer in the problem domain.